### DISCUSSION TITLE
GitHub errors are not logged as errors even though they are fatal when processing misconfigured GitHub repositories

### DISCUSSION THREAD

**Discussion by andersfischernielsen** [renovatebot/renovate#17857] [Suggest an Idea]
Renovate currently logs all errors using `logger.debug` instead of `logger.error`. This makes it hard to verify whether Renovate is processing repositories as expected (e.g. when self-hosting Renovate). These error messages currently drown in the debug output of Renovate.

Is this intended behaviour? Given that these errors show actual.... errors, I'd expect them to be logged as such 🤔 

---

I've opened [a PR](https://github.com/renovatebot/renovate/pull/17856) showing the locations where I'd expect an error to be logged 😄 Let me know what you think 🙏🏼

**Comment by rarkins** [comment]
I commented on your PR before seeing you'd raised this discussion, so copy/pasting my comment from there to here:

Some of these could be errors. Some could be warn. Some should remain as debug or info. Each needs to be evaluated individually before this PR can be approved.

Errors should be reserved for:

- Renovate code has a bug which needs to be fixed, or
- The operator of the bot needs to make a config or environmental change

In other words, they demand action to be taken.

Warns should be used for:
- Temporary errors which you may want to be aware of, but cannot directly take action about, or
- User errors, which is the responsibility of the user, not the bot operator

Even when there's a scenario where you the bot operator want to take responsibility for your users, that does not mean that logger.error is appropriate. For example consider myself watching the logs for half a million repos in the hosted app. If I care about every user error I'd never be able to notice the type I *can* address.

**Reply by andersfischernielsen** [reply]
I also responded in the PR 😄 We can continue here or there depending on your preference. 

My main question wrt. the PR revolves around the log level - if Renovate fails to process a repository it's (at least in our case) due to the repositories being misconfigured (wrt. branch protections, contributors, etc.) in the repository security settings. 

Would this then constitute a `warn`, since it's the responsibility of the user (in this case me, since I've set up a self-hosted Renovate instance and I need to either change the repository settings or notify the team in our organisation responsible for the repository with errors)?

**Reply by andersfischernielsen** [reply]
TLDR: My main concern is generally that it's currently very hard to detect when Renovate fails to process a repository. Since we love Renovate and use it a ton, I'd really like to somehow get a big alert when a team in our organisation suddenly isn't getting dependency updates due to repository misconfiguration/GitHub outages/etc.

**Reply by rarkins** [reply]
Can you annotate each of the log lines you wish to change describing which error/scenario would cause it? Then we can approve or modify each of them as we go

**Reply by andersfischernielsen** [reply]
I've just dug through our logs and found the commands that most often fail, requiring an action wrt. repo settings. 

We mainly see errors when Renovate attempts to push to branches and when Renovate tries to delete branches.

When pushing to branches where permissions aren't set up correctly we see output such as:

<details><summary>Token scope issues</summary>

```
Error: To https://github.com/pleo-io/<repo>.git
!	refs/heads/dependencies/actions-setup-node-3.x:refs/heads/dependencies/actions-setup-node-3.x	[remote rejected] (refusing to allow a Personal Access Token to create or update workflow `.github/workflows/check.yml` without `workflow` scope)
Done
Pushing to https://github.com/pleo-io/<repo>.git
POST git-receive-pack (730 bytes)
error: failed to push some refs to 'https://github.com/pleo-io/<repo>.git'
```
</details>

<details><summary>Lack of Renovate repo administrator permissions</summary>

```
Error: To https://github.com/pleo-io/<repo>.git
!	refs/heads/dependencies/chai-4.x:refs/heads/dependencies/chai-4.x	[remote rejected] (protected branch hook declined)
Done
Pushing to https://github.com/pleo-io/<repo>.git
POST git-receive-pack (899 bytes)
remote: error: GH006: Protected branch update failed for refs/heads/dependencies/chai-4.x.        
remote: error: Cannot force-push to this protected branch At least 1 approving review is required by reviewers with write access.        
error: failed to push some refs to 'https://github.com/pleo-io/<repo>.git'
```
</details>


<details><summary>Unable to delete branches due to lack of administrator permissions</summary>

```
Error: remote: error: GH006: Protected branch update failed for refs/heads/dependencies/analytics-node-3.x.        
remote: error: Cannot delete this protected branch        
To https://github.com/pleo-io/<repo>.git
 ! [remote rejected] dependencies/analytics-node-3.x (protected branch hook declined)
error: failed to push some refs to 'https://github.com/pleo-io/<repo>.git'
```
</details>

---

These all bubble up to the `Git function thrown` [error](https://github.com/pleo-oss/renovate/blob/fe6775d73f59d9362d664b62079871ba07884a32/lib/util/git/index.ts#L81), but are all related to pushing changes to the misconfigured repository.

The `git` commands behind these are roughly:

```
push origin <branch> --force-with-lease -u --no-verify --verbose --porcelain
```

```
push --delete origin <branch>
```

in turn showing that the [`deleteBranch`](https://github.com/pleo-oss/renovate/blob/fe6775d73f59d9362d664b62079871ba07884a32/lib/util/git/index.ts#L732) and [`pushCommit`](https://github.com/pleo-oss/renovate/blob/fe6775d73f59d9362d664b62079871ba07884a32/lib/util/git/index.ts#L1006) functions in `lib/util/git/index.ts` could be logged as `warn`s. 

As far as I can read, these both bubble up to the [`gitRetry`](https://github.com/pleo-oss/renovate/blob/fe6775d73f59d9362d664b62079871ba07884a32/lib/util/git/index.ts#L65) function in `lib/util/git/index.ts` where they are always logged as `debug`, which makes it harder to determine when to `warn` and when to `debug`. 

I'm unsure whether the thrown `Error` could be marked as an `ExternalHostError` to separate these "misconfigured repository" errors from normal errors or whether they should be marked in a different way, but they all indicate that some action needs to be taken on the repository side. I would at first glance consider the errors `ExternalHostErrors` since they _are_ caused by GitHub's branch protection or PAT scope checks, but how to then make sure that they are logged as `warn`s I don't know. 

Maybe the `checkForPlatformFailure` [function](https://github.com/pleo-oss/renovate/blob/1b206026829b39ff00e2dfcffc10986e2dadc7a0/lib/util/git/error.ts#L7) in `lib/util/git/error.ts` could be extended to log these errors as `warn`, since there's already logic in place to `error` log certain types of error. Up to you 😄 

Nonetheless, let me know what you think and whether a change is necessary at all 👍🏼

**Reply by viceice** [reply]
i think the check function needs to be updated with new error messages.

@rarkins some time ago we talked about inverting the error handling in git to only catch known errors and abort for any unexpected error instead of silently ignore unknown errors

**Reply by rarkins** [reply]
Doesn't gitRetry retry and then throw, unless the problem is a known external host error? Is there any scenario where it is "swallowing" the error?

**Reply by andersfischernielsen** [reply]
I just dug this out of DataDog (therefore, pardon the DD'ed formatting): 

```
Git function thrown

{
	[...]
	"content": {
		[...]
		"attributes": {
			"hostname": "renovate-xxxxxxxx-xxxxx",
			"repository": "pleo-io/<repo>",
			"err": {
				"message": "To https://github.com/pleo-io/<repo>.git
				!    refs/heads/dependencies/<dependency>:refs/heads/dependencies/<dependency>    [remote rejected] (refusing to allow a Personal Access Token to create or update workflow `.github/workflows/create-task-to-translate.yml` without `workflow` scope)
				Done
				Pushing to https://github.com/pleo-io/<repo>.git
				POST git-receive-pack (857 bytes)
				error: failed to push some refs to 'https://github.com/pleo-io/<repo>.git'
				",
				"task": {
					"commands": [
						"push",
						"origin",
						"dependencies/<dependency>:dependencies/<dependency>",
						"--force-with-lease",
						"-u",
						"--no-verify",
						"--verbose",
						"--porcelain"
					],
					"parser": "[function]",
					"format": "utf-8"
				},
				"stack": "Error: To https://github.com/pleo-io/<repo>.git
				!    refs/heads/dependencies/<dependency>:refs/heads/dependencies/<dependency>    [remote rejected] (refusing to allow a Personal Access Token to create or update workflow `.github/workflows/create-task-to-translate.yml` without `workflow` scope)
				Done
				Pushing to https://github.com/pleo-io/<repo>.git
				POST git-receive-pack (857 bytes)
				error: failed to push some refs to 'https://github.com/pleo-io/<repo>.git'

				    at Object.action (/usr/src/app/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)
				    at PluginStore.exec (/usr/src/app/node_modules/simple-git/src/lib/plugins/plugin-store.ts:29:29)
				    at /usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:127:42
				    at new Promise (<anonymous>)
				    at GitExecutorChain.handleTaskData (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:14)
				    at GitExecutorChain.<anonymous> (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:100:40)
				    at Generator.next (<anonymous>)
				    at fulfilled (/usr/src/app/node_modules/simple-git/dist/cjs/index.js:55:24)
				    at runMicrotasks (<anonymous>)
				    at processTicksAndRejections (internal/process/task_queues.js:95:5)"
			},
			"name": "renovate",
			"service": "renovate",
			"level": 20,
			"time": "2022-09-14T13:12:15.369Z",
			"pid": 8,
			"processed_log_level": "debug",
			"logContext": "XXXXXXXXXXXXXXXXXXXXX",
			"branch": "dependencies/<dependency>",
			"v": 0
		}
	}
}
```

If I read the [`bunyan` docs right](https://github.com/trentm/node-bunyan#levels), then `"level": 20` should be a `debug` log line. I can't see any part of the contents of the GH error in `checkForPlatformFailure`. The same applies to the other examples I added in the previous comment. Maybe I'm misreading the error or our DataDog is drunk 🤔 It does look as if it's being "swallowed" into a `debug` log line.

**Reply by rarkins** [reply]
This looks like it's a branch push failing? Do you know if it failed to create the branch, or failed to update the branch? Ideally you would be able to include the full logs for this branch when this error occurred (filter based on branchName and logContext)

**Reply by andersfischernielsen** [reply]
We've seen this error when Renovate fails to create the branch due to misconfigured repository permissions. Once we discover these, we've notified the respective repository owners and made sure that Renovate can do what it needs to do. We have to do this permissions dance every time the org adds a new repository, though 🤔 We can set up an alert based on the log contents, but I would just expect this to be a `warn` at first glance 🤷🏼

**Reply by rarkins** [reply]
Sorry didn't mean to let this drop. Fully support raising warns as necessary. 

If the git retry function is swallowing anything then we should fix that. But I thought it ultimately surfaced the errors. 

Let's start with some concrete examples.
